kimchi/src/prover.rs

```diff
@@ -419,79 +389,214 @@ where
-            assert_eq!(alphas_used, range::COMPLETE_ADD.len());
+            let alphas = all_alphas.get_powers(ConstraintType::Gate, 7);
+            let add_constraint = complete_add::constraint(alphas);
             let add4 = add_constraint.evaluations(&env);
             t4 += &add4;
+
+            if cfg!(test) {
+                let (_, res) = add4
+                    .clone()
+                    .interpolate()
+                    .divide_by_vanishing_poly(index.cs.domain.d1)
+                    .unwrap();
+                assert!(res.is_zero());
+            }
+
             drop(add4);
-            t4
-        };
 
-        // permutation
-        let (perm, bnd) =
-            index
+            // permutation
+            let alphas = all_alphas.get_alphas(ConstraintType::Permutation, 3);
+            let (perm, bnd) = index
                 .cs
-                .perm_quot(&lagrange, beta, gamma, &z_poly, &alphas[range::PERM])?;
-        let mut t8 = perm;
-        // scalar multiplication
-        let mul8 = varbasemul::constraint(range::MUL.start).evaluations(&env);
-        t8 += &mul8;
-        drop(mul8);
-        // endoscaling
-        let emul8 = endosclmul::constraint(2 + range::ENDML.start).evaluations(&env);
-        t8 += &emul8;
-        drop(emul8);
-        // endoscaling scalar computation
-        let emulscalar8 = endomul_scalar::constraint(range::ENDOMUL_SCALAR.start).evaluations(&env);
-        t8 += &emulscalar8;
-        drop(emulscalar8);
-        // poseidon
-        let pos8 = poseidon::constraint().evaluations(&env);
-        t8 += &pos8;
-        drop(pos8);
-
-        let t4 = match index.cs.chacha8.as_ref() {
-            None => t4,
-            Some(_) => {
-                let mut t4 = t4;
-                let chacha = chacha::constraint(range::CHACHA.start).evaluations(&env);
-                t4 += &chacha;
-                drop(chacha);
-                t4
+                .perm_quot(&lagrange, beta, gamma, &z_poly, alphas)?;
+            let mut t8 = perm;
+
+            if cfg!(test) {
+                let (_, res) = t8
+                    .clone()
+                    .interpolate()
+                    .divide_by_vanishing_poly(index.cs.domain.d1)
+                    .unwrap();
+                assert!(res.is_zero());
             }
-        };
 
-        // quotient polynomial for lookup
-        let (t4, t8) = match lookup_used {
-            None => (t4, t8),
-            Some(_) => combine_evaluations(
-                (t4, t8),
-                alpha,
-                alphas[alphas.len() - 1],
-                lookup::constraints(&index.cs.dummy_lookup_values[0], index.cs.domain.d1)
-                    .iter()
-                    .map(|e| e.evaluations(&env))
-                    .collect(),
-            ),
-        };
+            // scalar multiplication
+            let alphas = all_alphas.get_powers(ConstraintType::Gate, 21);
```

Comment:
`21` here is a 'magic constant', repeated here and in `index.rs`. It would be good to factor it out, as we had with range.

Reply:
Similarly with the others

Reply:
I have some ideas I want to explore on how to tackle this. I think it'd be nice if we didn't have to calculate this and we could directly get it from calling the linearization functions, but this should be in another PR.

Note that at the moment, if you enter a number too low or too high it will either crash due to the `zip_eq` (when combining) or because you didn't register enough. The only thing that can go wrong is that you register too many powers of alpha, which is not really a security issue.